pkg/controller/constrainttemplate/constrainttemplate_controller.go

```diff
@@ -848,3 +739,172 @@ func v1beta1ToV1(v1beta1Obj *admissionregistrationv1beta1.ValidatingAdmissionPol
+			err := r.reportErrorOnCTStatus(ctx, ErrCreateCode, "Could not get runtime VAP object", status, err)
+			return err
+		}
+
+		if err := controllerutil.SetControllerReference(ct, newVap, r.scheme); err != nil {
+			return err
+		}
+
+		if currentVap == nil {
+			logger.Info("creating VAP", "vapName", vapName)
+			if err := r.Create(ctx, newVap); err != nil {
+				logger.Info("creating VAP error", "vapName", vapName, "error", err)
+				err := r.reportErrorOnCTStatus(ctx, ErrCreateCode, "Could not create VAP object", status, err)
+				return err
+			}
+			// after VAP is created, trigger update event for all constraints
+			if err := r.triggerConstraintEvents(ctx, ct, status); err != nil {
+				return err
+			}
+		} else if !reflect.DeepEqual(currentVap, newVap) {
+			logger.Info("updating VAP")
+			if err := r.Update(ctx, newVap); err != nil {
+				err := r.reportErrorOnCTStatus(ctx, ErrUpdateCode, "Could not update VAP object", status, err)
+				return err
+			}
+		}
+	}
+	// do not generate VAP resources
+	// remove if exists
+	if !generateVap && isVapAPIEnabled && groupVersion != nil {
+		currentVap, err := vapForVersion(groupVersion)
+		if err != nil {
+			logger.Error(err, "error getting VAP object with respective groupVersion")
+			err := r.reportErrorOnCTStatus(ctx, ErrCreateCode, "Could not get VAP with correct group version", status, err)
+			return err
+		}
+		vapName := fmt.Sprintf("gatekeeper-%s", unversionedCT.GetName())
+		logger.Info("check if VAP exists", "vapName", vapName)
+		if err := r.Get(ctx, types.NamespacedName{Name: vapName}, currentVap); err != nil {
+			if !apierrors.IsNotFound(err) && !errors.As(err, &discoveryErr) && !meta.IsNoMatchError(err) {
+				return err
+			}
+			currentVap = nil
+		}
+		if currentVap != nil {
+			logger.Info("deleting VAP")
+			if err := r.Delete(ctx, currentVap); err != nil {
+				err := r.reportErrorOnCTStatus(ctx, ErrUpdateCode, "Could not delete VAP object", status, err)
+				return err
+			}
+			// after VAP is deleted, trigger update event for all constraints
+			if err := r.triggerConstraintEvents(ctx, ct, status); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileConstraintTemplate) updateTemplateWithBlockVAPBGenerationAnnotations(ctx context.Context, ct *v1beta1.ConstraintTemplate) error {
+	currentTime := time.Now()
+	switch {
+	case ct.Annotations == nil || ct.Annotations[constraint.BlockVAPBGenerationUntilAnnotation] == "":
+		if ct.Annotations == nil {
+			ct.Annotations = make(map[string]string)
+		}
+		ct.Annotations[constraint.BlockVAPBGenerationUntilAnnotation] = currentTime.Add(time.Duration(*defaultWaitForVAPBGeneration) * time.Second).Format(time.RFC3339)
+	default:
+		until := ct.Annotations[constraint.BlockVAPBGenerationUntilAnnotation]
+		t, err := time.Parse(time.RFC3339, until)
+		if err != nil {
+			return err
+		}
+		if t.Before(currentTime.Add(time.Duration(*defaultWaitForVAPBGeneration*2) * time.Second)) {
```

Comment:
why is this `* 2`? can you add a comment for this?

Reply:
I modified this to remove *2, added the comment as well.